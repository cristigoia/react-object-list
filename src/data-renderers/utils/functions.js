/**
 * Gets the value of the cell from the row and keys provided
 * @param  {Object} row  Row supplied
 * @param  {Array|string} keys List of keys used to access the value
 * @return {(Object|string|number)}      Value from row, or row itself if no keys
 */
const getValueFromAccessor = (row, keys) => {
  if (!Array.isArray(keys)) keys = keys.split('.')
  let value = row
  const cloneKeys = [...keys]
  while (value && cloneKeys.length) {
    const thisKey = cloneKeys.shift()
    if (thisKey in value) {
      value = value[thisKey]
    }
  }
  return value
}

export {
  getValueFromAccessor,
}
