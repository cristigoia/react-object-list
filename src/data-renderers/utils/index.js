export {
  getValueFromAccessor,
  handleRowClick,
} from './functions'
