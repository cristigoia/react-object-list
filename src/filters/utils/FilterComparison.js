import React from 'react'
import PropTypes from 'prop-types'
import Select from 'react-select'

class FilterComparison extends React.Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string,
      value: Select.propTypes.value,
    })),
    value: Select.propTypes.value,
    onChange: PropTypes.func.isRequired,
  }

  render() {
    const {options, value} = this.props
    return (
      <div className="objectlist-current-filter__filter-comparison">
        <Select
          options={options}
          value={value}
          onChange={newValue => this.props.onChange(newValue.value)}
          clearable={false}
        />
      </div>
    )
  }
}

export default FilterComparison
