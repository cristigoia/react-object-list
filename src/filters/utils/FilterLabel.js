import React from 'react'
import PropTypes from 'prop-types'

/**
 * Component to render label for filters
 */
class FilterLabel extends React.Component {
  static propTypes = {
    /** Text to render in the label */
    label: PropTypes.string.isRequired,
  }
  render() {
    return (
      <label className="objectlist-current-filter__filter-name">{this.props.label}:</label>
    )
  }
}

export default FilterLabel
