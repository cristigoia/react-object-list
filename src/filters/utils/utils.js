const possibleQueryStringLookupExpressions = ['exact', 'lt', 'gt', 'gte', 'lte']

export {
  possibleQueryStringLookupExpressions,
}
