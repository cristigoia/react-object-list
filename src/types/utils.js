function floatPrecision(floatValue, precision) {
  floatValue = parseFloat(floatValue)
  if (isNaN(floatValue)) { return parseFloat('0').toFixed(precision) } else {
    const power = Math.pow(10, precision)
    floatValue = (Math.round(floatValue * power) / power).toFixed(precision)
    return floatValue.toString()
  }
}

function commaNumber(numberString) {
  const parts = numberString.toString().split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',')
  return parts.join('.')
}
/**
 * Converts a float to stringed currency representation
 * @param  {number} floatNumber Amount to convert
 * @return {string}             Dollar string
 */
export function currency(floatNumber) {
  return `${floatNumber < 0 ? '-' : ''}$${commaNumber(floatPrecision(Math.abs(floatNumber), 2))}`
}
