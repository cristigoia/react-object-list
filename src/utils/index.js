export * from './constants'
export * from './proptypes'
export * from './functions'
export loadingSpinner from './loading'
