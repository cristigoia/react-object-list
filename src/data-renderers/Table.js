import React, { Component } from 'react'
import PropTypes from 'prop-types'

import TableHeader from './TableHeader'
import {AllSelector, Selector} from '../types'
import { getVisibleColumns, setColumnLabels } from '../utils/functions'
import { DATA_TYPE } from '../utils/constants'

export default class TableRenderer extends Component {
  static propTypes = {
    columns: PropTypes.arrayOf(PropTypes.oneOfType([DATA_TYPE, PropTypes.arrayOf(DATA_TYPE)])),
    meta: PropTypes.object,
    saveColumnWidth: PropTypes.func,
    data: PropTypes.array,
    /** eg. { label: {width: "100"} } where 'label' is the label generated for the column */
    columnWidths: PropTypes.object,
    /** set the sort to the given sortkeys */
    updateSorting: PropTypes.func,
    /** IDs of currently selected items set to true if selected */
    selection: PropTypes.object,
    /** Function to select one */
    select: PropTypes.func,
  }

  static defaultProps = {
    meta: {
      extraColumns: [],
      sortKeys: [],
    },
    columnWidths: {},
    selection: {},
  }

  state = {
    columns: getVisibleColumns(setColumnLabels(this.props.columns), this.props.meta.extraColumns),
  }

  componentWillReceiveProps(nextProps) {
    // TODO: optimise performance and only do this if the columns prop has changed
    this.setState(() => ({columns: getVisibleColumns(setColumnLabels(nextProps.columns), nextProps.meta.extraColumns)})) 
  }

  _renderHeaderRowHelper = () => {
    return this.state.columns.map(column => {
      const label = Array.isArray(column) ? column[0].label : column.label
      const width = this.props.columnWidths[label] ? this.props.columnWidths[label].width : null
      return (
        <TableHeader
          key={`header-${label}`}
          headerItems={column}
          label={label}
          width={width || column.width}
          sortKeys={this.props.meta.sortKeys}
          saveWidth={this.props.saveColumnWidth}
          updateSorting={this.props.updateSorting}
        />
      )
    })
  }

  _renderItemRowsHelper = () => {
    const {data, selection, select} = this.props
    return data.map((row, rowIndex) => {
      const selected = row.id in selection
      return (
        <tr key={`row-${rowIndex}`} className="apilist-table__row">
          {select && (
            <td className="apilist-table__td">
              <Selector
                toggleSelect={select}
                selected={selected}
                id={row.id}
              />
            </td>
          )}
          {this.state.columns.map((column, cellIndex) => {
            const toRender = Array.isArray(column) ? column : [column]
            return (
              <td key={`${rowIndex}-${cellIndex}`} className="apilist-table__td">
                {toRender.map(RenderItem => {
                  if (RenderItem.item) {
                    return RenderItem.item({
                      row: row,
                      column: RenderItem,
                      value: row[RenderItem.dataKey],
                    })
                  }
                  return row[RenderItem.dataKey]
                }
                ).join(' ')}
              </td>
            )
          })}
        </tr>
      )
    })
  }

  render() {
    const {selection, data, select} = this.props
    return (
      <div className="apilist-table--scroll">
        <table className="apilist-table">
          <thead>
            <tr>
              {select && (
                <th className="apilist-table__th apilist-table__th--selector">
                  <AllSelector
                    allSelected={Object.keys(selection).length >= data.length}
                    toggleSelectAll={() => select(data.map(row => row.id))}
                  />
                </th>
              )}
              {this._renderHeaderRowHelper()}
            </tr>
          </thead>
          <tbody>
            {this._renderItemRowsHelper()}
          </tbody>
        </table>
      </div>
    )
  }
}
