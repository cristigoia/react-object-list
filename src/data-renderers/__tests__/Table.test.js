import React from 'react'
import { snapshotTest } from 'utils/tests'
import Table from '../Table'

jest.mock('../TableHeader', () => 'TableHeader')

describe('Table', () => {
  describe('Snapshots', () => {
    const props = {
      columns: [
        {
          dataKey: 'name',
          header: 'Name',
          item: ({row: {attributes: {name}}}) => name,
          optional: false,
          sortable: true,
        },
        [{
          dataKey: 'age',
          header: 'Age',
          item: ({row: {attributes: {age}}}) => (<span>{age}</span>),
          optional: false,
          sortable: true,
        },
        {
          dataKey: 'gender',
          header: 'Gender',
          item: ({row: {attributes: {gender}}}) => gender,
          optional: false,
          sortable: true,
        }],
      ],
      columnWidths: {'age_gender': {width: 50}},
      saveColumnWidth: jest.fn(),
      data: [
        {type: 'Person', attributes: {name: 'Sam', age: '25', gender: 'M'}},
        {type: 'Person', attributes: {name: 'Mary', age: '12', gender: 'F'}},
        {type: 'Person', attributes: {name: 'Q', age: '3', gender: '?'}},
        {type: 'Person', attributes: {name: 'Peter', age: '111', gender: 'FM'}},
        {type: 'Person', attributes: {name: 'Amy', age: '32', gender: 'Confused'}},
      ],
    }

    it('renders correctly', () => {
      snapshotTest(<Table {...props} />)
    })
  })
})
