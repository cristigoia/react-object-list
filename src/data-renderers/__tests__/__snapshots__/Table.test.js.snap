// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Snapshots renders correctly 1`] = `
<div
  className="apilist-table--scroll"
>
  <table
    className="apilist-table"
  >
    <thead>
      <tr>
        <TableHeader
          headerItems={
            Object {
              "dataKey": "name",
              "header": "Name",
              "item": [Function],
              "label": "name",
              "optional": false,
              "sortable": true,
            }
          }
          label="name"
          saveWidth={[MockFunction]}
          sortKeys={Array []}
          updateSorting={undefined}
          width={undefined}
        />
        <TableHeader
          headerItems={
            Array [
              Object {
                "dataKey": "age",
                "header": "Age",
                "item": [Function],
                "label": "age_gender",
                "optional": false,
                "sortable": true,
              },
              Object {
                "dataKey": "gender",
                "header": "Gender",
                "item": [Function],
                "label": "age_gender",
                "optional": false,
                "sortable": true,
              },
            ]
          }
          label="age_gender"
          saveWidth={[MockFunction]}
          sortKeys={Array []}
          updateSorting={undefined}
          width={50}
        />
      </tr>
    </thead>
    <tbody>
      <tr
        className="apilist-table__row"
      >
        <td
          className="apilist-table__td"
        >
          Sam
        </td>
        <td
          className="apilist-table__td"
        >
          [object Object] M
        </td>
      </tr>
      <tr
        className="apilist-table__row"
      >
        <td
          className="apilist-table__td"
        >
          Mary
        </td>
        <td
          className="apilist-table__td"
        >
          [object Object] F
        </td>
      </tr>
      <tr
        className="apilist-table__row"
      >
        <td
          className="apilist-table__td"
        >
          Q
        </td>
        <td
          className="apilist-table__td"
        >
          [object Object] ?
        </td>
      </tr>
      <tr
        className="apilist-table__row"
      >
        <td
          className="apilist-table__td"
        >
          Peter
        </td>
        <td
          className="apilist-table__td"
        >
          [object Object] FM
        </td>
      </tr>
      <tr
        className="apilist-table__row"
      >
        <td
          className="apilist-table__td"
        >
          Amy
        </td>
        <td
          className="apilist-table__td"
        >
          [object Object] Confused
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
